lib/puppetserver/ca/certificate_authority.rb

```diff
@@ -0,0 +1,224 @@
+require 'puppetserver/ca/utils/http_client'
+
+require 'json'
+
+module Puppetserver
+  module Ca
+    class CertificateAuthority
+
+      include Puppetserver::Ca::Utils
+
+      REVOKE_BODY = JSON.dump({ desired_state: 'revoked' })
+      SIGN_BODY =   JSON.dump({ desired_state: 'signed' })
+
+      def initialize(logger, settings)
+        @logger = logger
+        @client = HttpClient.new(settings)
+        @ca_server = settings[:ca_server]
+        @ca_port = settings[:ca_port]
+      end
+
+      # Returns a URI-like wrapper around CA specific urls
+      def make_ca_url(resource_type = nil, certname = nil)
+        HttpClient::URL.new('https', @ca_server, @ca_port, 'puppet-ca', 'v1', resource_type, certname)
+      end
+
+      def sign_certs(certnames)
+        put(certnames,
+            resource_type: 'certificate_status',
+            body: SIGN_BODY,
+            type: :sign)
+      end
+
+      def revoke_certs(certnames)
+        put(certnames,
+            resource_type: 'certificate_status',
+            body: REVOKE_BODY,
+            type: :revoke)
+      end
+
+      def submit_certificate_request(certname, csr)
+        put([certname],
+            resource_type: 'certificate_request',
+            body: csr.to_pem,
+            headers: {'Content-Type' => 'text/plain'},
+            type: :submit)
+      end
+
+      # Make an HTTP PUT request to CA
+      # @param resource_type [String] the resource type of url
+      # @param certnames [Array] array of certnames
+      # @param body [JSON/String] body of the put request
+      # @param type [Symbol] type of error processing to perform on result
+      # @return [Boolean] whether all requests were successful
+      def put(certnames, resource_type:, body:, type:, headers: nil)
```

Comment:
`headers` needs to default to the empty hash. If it's nil, when it gets passed to `DEFAULT_HEADERS.merge(headers)` inside the HTTPClient, it will crash.